### ISSUE TITLE
Reuse `/run/current-system/sw` instead of `/run/system-manager`

### ISSUE THREAD

**Issue by bryango** [numtide/system-manager#45]
**Is your feature request related to a problem? Please describe.**

I was trying to adapt nixos modules for system-manager and noticed that many of them have hard-coded `/run/current-system/sw`; see e.g.
- https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%2Frun%2Fcurrent-system%2Fsw+NOT+path%3A%2F%5Enixos%5C%2Ftests%5C%2F%2F&type=code

One must modify `/run/current-system -> /run/system-manager` in order to port such modules.

**Describe the solution you'd like**

Since `/run/current-system` is not used in a generic non-NixOS system I believe we can simply re-use the path.

**Describe alternatives you've considered**

- Rename each `/run/current-system -> /run/system-manager` module by module.
- Alternatively, provide an option to symlink `/run/current-system -> /run/system-manager` when profiles are activated.

**Additional context**

I am not sure if it's safe to do this. It feels safe to me since `system-manager` should only be used on non-NixOS so there will not be conflicts. But I don't know.

**Comment by R-VdP** [comment]
I'd prefer doing the symlink and making it optional, so that it can be turned off in case anyone has issues with it.

**Comment by bryango** [comment]
Oh I just noticed that NixOS itself is on the list of supported systems:
https://github.com/numtide/system-manager/blob/549bc383395f504261ef4f36fa26d04c37809d12/nix/modules/default.nix#L139-L145
So indeed it should be an option that one can toggle on or off.

**Comment by r-vdp** [comment]
I don't think we have that right now, but it would be rather easy to introduce an option for this. It might be a rather hard sell to get that merged into nixpkgs though.

**Comment by aanderse** [comment]
> It might be a rather hard sell to get that merged into nixpkgs though.

i suggest there is [prior art](https://github.com/NixOS/nixpkgs/blob/e33debd24980fb317642cbf26fa492cebeb7789a/nixos/modules/security/wrappers/default.nix#L218) for a change like this

additionally, with [more](https://github.com/nix-community/NixNG) and [more](https://github.com/nixos-bsd/nixbsd) nix based systems popping up i think this is a good idea to have an option for this at least as an `internal` option

tangentially related... seems like a PR replacing all instances of `/run/wrappers/bin` in `nixpkgs/nixos` with `config.security.wrapperDir` is a good idea... we might even consider setting `security.wrapperDir = "/sbin";` on `system-manager` :man_shrugging:

---

> To help make modules more portable, is there some attribute that would return `/run/current-system/sw` in a nixosModule and `/run/system-manager` on a systemModule? I’d be happy to fix the hardcoding of /run/current-system/sw in nixpkgs if so.

please do, and if you don't mind i would love a ping on that PR

**Comment by Meliketoaste** [comment]
In nixpkgs they also use `environment.pathsToLink` this seems relevant but Im dumb.
```nix
pathsToLink = lib.mkOption {
        type = lib.types.listOf lib.types.str;
        # Note: We need `/lib' to be among `pathsToLink' for NSS modules
        # to work.
        default = [];
        example = ["/"];
        description = "List of directories to be symlinked in {file}`/run/current-system/sw`.";
};
```

**Comment by sellout** [comment]
> tangentially related... seems like a PR replacing all instances of `/run/wrappers/bin` in `nixpkgs/nixos` with `config.security.wrapperDir` is a good idea... we might even consider setting `security.wrapperDir = "/sbin";` on `system-manager` 🤷‍♂️
> 
> > To help make modules more portable, is there some attribute that would return `/run/current-system/sw` in a nixosModule and `/run/system-manager` on a systemModule? I’d be happy to fix the hardcoding of /run/current-system/sw in nixpkgs if so.
> 
> please do, and if you don't mind i would love a ping on that PR

There’s no PR yet, but I have a branch that at least makes progress on both of these changes (it’s old, but I just rebased it, and figured it would be good if someone knew that it exists): https://github.com/sellout/nixpkgs/tree/config-paths

Mostly a big find/replace, the hurdles that exist are

1. Non-NixOS pieces (e.g., packages) that expect these exact paths to exist. Some take a `wrapperDir` arg, but they default to a hardcoded `"/run/wrappers/bin"` – I’m thinking they should _all_ take a `wrapperDir` arg and the default should be removed to be set more globally (and a corresponding change for the new `environment.systemDir` option)
2. Non-Nix files (e.g., patches) that hardcode these paths. The quick solution here might be an `assert` that ensures `wrapperDir` (and/or `systemDir`) hasn’t been changed from the default. Over time, the hardcoded values can be replaced with other substitution approaches in nix files that have access to `wrapperDir`.

It would also be nice if Nixpkgs had some CI auditing that could check if new instances of the hardcoded paths appear in PRs. (Does it?)
